File: src/main/java/de/iisys/drossner/algodat/sort/MergeSort.java
package de.iisys.drossner.algodat.sort;

import java.util.Arrays;
import java.util.Random;

public class MergeSort {

    public static void main(String[] args) {
        int[] array = new Random().ints(100, 0, 1000).toArray();
        Arrays.stream(array).forEach(System.out::println);
        mergesort(array);
        System.out.println("============ Sortiert ===========");
        Arrays.stream(array).forEach(System.out::println);
    }

    public static void mergesort(int[] array){
        mergesort(array, 0, array.length);
    }

    private static void mergesort(int[] array, int start, int end) {
        int mid = (start + end) / 2;

        if(start < end - 1) {
            mergesort(array, start, mid);
            mergesort(array, mid, end);
            merge(array, start, mid, end);
        }
    }



    private static void merge(int[] array, int start, int mid, int end){
        int[] left = Arrays.copyOfRange(array, start, mid);
        int[] right = Arrays.copyOfRange(array, mid, end);

        int leftCursor = 0;
        int rightCursor = 0;

        for(int i = start; i < end ; i++) {
            int lowLeft = (leftCursor < left.length)? left[leftCursor] : Integer.MAX_VALUE;
            int lowRight = (rightCursor < right.length)? right[rightCursor] : Integer.MAX_VALUE;
            if(lowLeft <= lowRight) {
                array[i] = lowLeft;
                leftCursor++;
            } else {
                array[i] = lowRight;
                rightCursor++;
            }
        }
    }
}

File: src/main/kotlin/de/iisys/drossner/algodat/sort/MergeSort.kt
package de.iisys.drossner.algodat.sort

import java.util.*

fun mergesort(a: IntArray, start: Int, end: Int){
    if (start < end - 1) {
        val mid = (start + end) / 2
        mergesort(a, start, mid)
        mergesort(a, mid, end)
        merge(a, start, mid, end)
    }
}

fun merge(a: IntArray, start: Int, mid: Int, end: Int){
    val tempa = IntArray(mid - start, {index -> a[index + start]})
    val tempb = IntArray(end - mid, {index -> a[index + mid]})

    var aIndex = 0
    var bIndex = 0

    for (k in start until end){
        a[k] =
                if (tempa.getOrElse(aIndex, {Int.MAX_VALUE}) <= tempb.getOrElse(bIndex, {Int.MAX_VALUE})){
                    tempa[aIndex++]
                } else {
                    tempb[bIndex++]
                }
    }
}

fun main(args: Array<String>) {
    val test = intArrayOf(10, 1, 18, 86, 12, 3, 8, 81, 22, 13, 4, 20, 21, 36)
    mergesort(test, 0, test.size)
    println(Arrays.toString(test))
}